jps/model-impl/src/com/intellij/openapi/fileTypes/impl/FileTypeAssocTable.java

```diff
@@ -42,7 +51,7 @@ private FileTypeAssocTable(@NotNull Map<? extends CharSequence, ? extends T> ext
     myExactFileNameAnyCaseMappings = concurrentCharSequenceMapBuilder.build(exactFileNameAnyCaseMappings, false);
     myConcurrentCharSequenceMapBuilder = concurrentCharSequenceMapBuilder;
     myHashBangMap = new ConcurrentHashMap<>(hashBangMap);
-    myMatchingMappings = ContainerUtil.createLockFreeCopyOnWriteList(matchingMappings);
+    myMatchingMappings = new CopyOnWriteArrayList<>(matchingMappings);
```

Comment:
please revert

Reply:
TL;DR The JB custom lock-free COW List is not sortable, which is why I reverted to the built-in Java one, which is sortable. If you want to stay with the JB one, I'm not quite sure how you want me to implement that? Implement atomic sorting for JB's COW arraylist as a part of this ticket?

For a longer explanation, here's the full context:

Since I didn't have any direct line of contact to you to ask questions about the implementation, I went with my best effort of understanding it from the code and relevant tickets. So I'll take this opportunity to explain it and ask questions, hope you don't mind.

So yes, while investigating this I've noticed that the code first was not thread-safe, then made thread-safe with `Collections.synchronizedList` (there was no relevant ticket linked, but I assume that's a self-evident issue), then later to Java's built-in COW arraylist (the ticket does not explain why, I'd assume it would be contention issues?) and then finally a custom JB COW arraylist implementation was made by you, as a part of this ticket — https://youtrack.jetbrains.com/issue/KTIJ-27475.

I figured the cleanest solution would've been a sorted datastructure (such as a sorted map), but given it would require benchmarking and none of the changes in the place had any test data to benchmark with, I decided against that.

So then next best idea was to just sort on insertion — it's kind of crude, but maybe modifications don't happen often enough for it to matter. But it turns out that JB COW arraylist is not sortable — Java's sort requires a mutable iterator by default (c.f. https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/List.java#L505-L513) while COW arraylist (both stdlib's and JB's) have immutable iterators. The stdlib list works around that by defining it's own `sort` implenetation (c.f. https://github.com/openjdk/jdk/blob/jdk-17%2B35/src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java#L886-L898) but JB's list doesn't. This makes sorting fail using JB's COW arraylist.

So the next best thing I could think of without having any direct line of contact to you is to try reverting to stdlib's COW arraylist and try benchmarking if it makes any difference. Your PR for KTIJ-27475 makes it look like the introduction of JB's COW arraylist wasn't really related to `FileTypeAssocTable.java`, but some other issue (hard to tell for sure though, without more context that's unavailable to me) so I figured it's worth a try.

So this is the implementation I ended up with and I have benchmarked it by building the IDE with and without that change and using those built IDEs to compile IntelliJ again, with profiler enabled. In both cases full IntelliJ compilation took ~12min 30s and churned about ~50GB of memory and seemed to behave similarly. So I figured it's okay to revert that and use stdlib's COW arraylist in this case.

If you have any projects I can test this that this code behaved pathologically with stdlib's implementation, I would appreciate being able to test with them to see if I'm not breaking anything here. Otherwise it's hard to see how this change has a negative impact. It's quite possible I might be missing something, but I'm not sure what.

Reply:
Hi. It's a valid point about ContainerUtil.createLockFreeCopyOnWriteList() not being (atomically) sortable, thanks for noticing. I've pushed a version with overridden sort() method, so now it's sortable. About testing the difference: it's hard to spot any differencein performance of ContainerUtil.createLockFreeCopyOnWriteList(), off-the-shelf COWAL or synchronizedLIst here, since this code is not on any critical path. I'm advocating  ContainerUtil.createLockFreeCopyOnWriteList here for consistency/simplicity reasons, not the performance.

Reply:
Makes sense, now that those operations were added it should be okay to use the JB COW arraylist variant and drop the locks. I'll update the PR accordingly.